File: android/app/src/cn/java/com/app/wechat/ESPWeChatModule.kt
/*
 * SPDX-FileCopyrightText: 2026 Espressif Systems (Shanghai) CO LTD
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package com.app.wechat

import android.os.Handler
import android.os.Looper
import android.util.Log
import com.app.BuildConfig
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.tencent.mm.opensdk.modelmsg.SendAuth
import com.tencent.mm.opensdk.openapi.IWXAPI
import com.tencent.mm.opensdk.openapi.WXAPIFactory

/**
 * ESPWeChatModule — bridge between the React Native JS layer and the native
 * WeChat SDK. CN-flavor only (the WeChat SDK ships solely in the CN build).
 *
 * Flow:
 *   init():                register the app with WeChat (once, on module load)
 *   initiateWeChatLogin(): send the auth request when the user taps the button
 *   WXEntryActivity:       receive the auth code and resolve the pending JS
 *                          promise via this module; the RainMaker Base SDK then
 *                          exchanges the code for tokens on the JS side
 *
 * The App ID comes from `BuildConfig.WECHAT_APP_ID` (synced from `WECHAT_APP_ID`
 * in .env).
 *
 * NOTE: registerApp and sendReq must not run in the same frame — the SDK commits
 * registration to an IPC store that sendReq reads, so we register early (init)
 * and only sendReq on tap.
 */
class ESPWeChatModule(reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    companion object {
        private const val TAG = "ESPWeChatModule"

        /** WeChat App ID for this build (empty when unconfigured). */
        @JvmField
        val WECHAT_APP_ID: String = BuildConfig.WECHAT_APP_ID

        @JvmField
        @Volatile
        var instance: ESPWeChatModule? = null
    }

    private val mainHandler = Handler(Looper.getMainLooper())

    @Volatile
    private var pendingLoginPromise: Promise? = null

    @Volatile
    var currentWxApi: IWXAPI? = null

    init {
        instance = this
        // Register with WeChat immediately so the SDK has committed registration
        // to its IPC store well before sendReq runs (on button tap).
        val appId = WECHAT_APP_ID
        if (appId.isNotBlank()) {
            mainHandler.post {
                try {
                    val api = WXAPIFactory.createWXAPI(reactContext, appId, true)
                    api.registerApp(appId)
                    currentWxApi = api
                    Log.d(TAG, "init: registered appId=$appId, isInstalled=${api.isWXAppInstalled}")
                } catch (e: Exception) {
                    Log.e(TAG, "init: registerApp failed: ${e.message}", e)
                }
            }
        } else {
            Log.w(TAG, "init: WECHAT_APP_ID is blank in BuildConfig, skipping registration")
        }
    }

    override fun getName() = "ESPWeChatModule"

    /**
     * WeChat login entry point from JS. Resolves with `{ code }` — the raw
     * WeChat authorization code; the RainMaker Base SDK performs the token
     * exchange on the JS side. The `tokenUrl` / `clientId` / `redirectUri`
     * parameters are unused but kept so the bridge method signature stays
     * stable across platforms.
     */
    @Suppress("UNUSED_PARAMETER")
    @ReactMethod
    fun initiateWeChatLogin(
        tokenUrl: String,
        clientId: String,
        redirectUri: String,
        promise: Promise
    ) {
        val appId = WECHAT_APP_ID
        if (appId.isBlank()) {
            promise.reject("WECHAT_CONFIG_ERROR", "WeChat App ID is not configured in this build")
            return
        }

        pendingLoginPromise = promise

        mainHandler.post {
            try {
                var api = currentWxApi
                if (api == null) {
                    val ctx = reactApplicationContext.currentActivity ?: reactApplicationContext
                    api = WXAPIFactory.createWXAPI(ctx, appId, true)
                    api.registerApp(appId)
                    currentWxApi = api
                    // Defer sendReq one frame so registration commits first.
                    mainHandler.post { sendAuthRequest(api) }
                    return@post
                }

                if (!api.isWXAppInstalled) {
                    rejectLogin("WECHAT_NOT_INSTALLED: WeChat is not installed")
                    return@post
                }

                sendAuthRequest(api)
            } catch (e: Exception) {
                Log.e(TAG, "initiateWeChatLogin error: ${e.message}", e)
                rejectLogin("WECHAT_AUTH_FAILED: ${e.message}")
            }
        }
    }

    private fun sendAuthRequest(api: IWXAPI) {
        val req = SendAuth.Req()
        req.scope = "snsapi_userinfo"
        req.state = "esp_rainmaker_wechat_login"
        val success = api.sendReq(req)
        Log.d(TAG, "sendAuthReq dispatched: $success")
        if (!success) {
            rejectLogin("WECHAT_AUTH_FAILED: sendReq returned false")
        }
    }

    /**
     * Called by WXEntryActivity with the raw WeChat auth code. JS then hands the
     * code to the Base SDK, which performs the token exchange. This is the
     * current path — token exchange lives in the SDK, not the app.
     */
    fun resolveWithCode(code: String) {
        val map = Arguments.createMap().apply {
            putString("code", code)
        }
        pendingLoginPromise?.resolve(map)
        pendingLoginPromise = null
    }

    /** Called by WXEntryActivity (or internally) when the attempt fails/cancels. */
    fun rejectLogin(error: String) {
        Log.e(TAG, "rejectLogin: $error")
        pendingLoginPromise?.reject("WECHAT_AUTH_FAILED", error)
        pendingLoginPromise = null
    }

    @Suppress("DEPRECATION")
    override fun onCatalystInstanceDestroy() {
        super.onCatalystInstanceDestroy()
        pendingLoginPromise?.reject("WECHAT_AUTH_FAILED", "Module destroyed")
        pendingLoginPromise = null
        currentWxApi = null
        instance = null
    }
}

File: ios/WeChat/ESPWeChatModule.swift
/*
 * SPDX-FileCopyrightText: 2026 Espressif Systems (Shanghai) CO LTD
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import Foundation
import UIKit
import React

/**
 * ESPWeChatModule — bridge between the React Native JS layer and the native
 * WeChat SDK (CN region only).
 *
 * JS calls `initiateWeChatLogin`; this module sends the WeChat auth request.
 * `AppDelegate` (the `WXApiDelegate`) forwards the auth response to
 * `handleAuthResponse`, which resolves the pending JS promise with `{ code }` —
 * the raw WeChat authorization code. The RainMaker Base SDK then exchanges the
 * code for tokens on the JS side (`loginWithOauthCode`).
 *
 * The WeChat App ID / Universal Link are registered by `AppDelegate` at launch
 * (read from Info.plist: `WeChatAppID` / `WeChatUniversalLink`).
 */
@objc(ESPWeChatModule)
class ESPWeChatModule: NSObject, RCTBridgeModule {

  @objc static var shared: ESPWeChatModule?

  private var pendingResolve: RCTPromiseResolveBlock?
  private var pendingReject: RCTPromiseRejectBlock?

  static func moduleName() -> String { "ESPWeChatModule" }

  @objc static func requiresMainQueueSetup() -> Bool { false }

  override init() {
    super.init()
    ESPWeChatModule.shared = self
  }

  // MARK: - JS entry point

  /// Sends the WeChat auth request and later resolves with `{ code }`. App ID
  /// comes from Info.plist (registered at launch by AppDelegate). The
  /// `tokenUrl` / `clientId` / `redirectUri` parameters are unused — the Base
  /// SDK performs the token exchange on the JS side — but are kept so the
  /// bridge method signature stays stable across platforms.
  @objc(initiateWeChatLogin:clientId:redirectUri:resolver:rejecter:)
  func initiateWeChatLogin(_ tokenUrl: String,
                           clientId: String,
                           redirectUri: String,
                           resolver resolve: @escaping RCTPromiseResolveBlock,
                           rejecter reject: @escaping RCTPromiseRejectBlock) {

    let appId = Bundle.main.infoDictionary?["WeChatAppID"] as? String ?? ""
    if appId.isEmpty {
      reject("WECHAT_CONFIG_ERROR", "WeChatAppID not found in Info.plist", nil)
      return
    }

    pendingResolve = resolve
    pendingReject = reject

    DispatchQueue.main.async { [weak self] in
      guard WXApi.isWXAppInstalled() else {
        self?.finishReject("WECHAT_NOT_INSTALLED", "WeChat is not installed on this device")
        return
      }

      let req = SendAuthReq()
      req.scope = "snsapi_userinfo"
      req.state = "esp_rainmaker_wechat_login"

      guard let delegate = UIApplication.shared.delegate as? WXApiDelegate,
            let topVC = ESPWeChatModule.topViewController() else {
        self?.finishReject("WECHAT_AUTH_FAILED", "No view controller available for WeChat auth")
        return
      }

      WXApi.sendAuthReq(req, viewController: topVC, delegate: delegate) { success in
        if !success {
          self?.finishReject("WECHAT_AUTH_FAILED", "Failed to send WeChat auth request")
        }
      }
    }
  }

  // MARK: - Auth response (forwarded by AppDelegate)

  @objc func handleAuthResponse(code: String?, errCode: Int32, errStr: String?) {
    guard pendingResolve != nil, pendingReject != nil else { return }

    switch errCode {
    case 0:
      if let authCode = code {
        finishResolve(["code": authCode])
      } else {
        finishReject("WECHAT_AUTH_FAILED", "Auth succeeded but no code received")
      }
    case -2:
      finishReject("WECHAT_CANCELLED", "WeChat login was cancelled by user")
    case -4:
      finishReject("WECHAT_AUTH_DENIED", "WeChat login was denied by user")
    default:
      finishReject("WECHAT_AUTH_FAILED", "errCode=\(errCode), errStr=\(errStr ?? "unknown")")
    }
  }

  // MARK: - Helpers

  /// Resolves the front-most view controller (used by the WeChat SDK as the
  /// presenter for the not-installed QR fallback).
  private static func topViewController() -> UIViewController? {
    let keyWindow = UIApplication.shared.connectedScenes
      .compactMap { $0 as? UIWindowScene }
      .flatMap { $0.windows }
      .first { $0.isKeyWindow }

    var top = keyWindow?.rootViewController
    while let presented = top?.presentedViewController {
      top = presented
    }
    return top
  }

  // MARK: - Promise helpers (clear pending state atomically)

  private func finishResolve(_ value: Any) {
    let resolve = pendingResolve
    pendingResolve = nil
    pendingReject = nil
    resolve?(value)
  }

  private func finishReject(_ code: String, _ message: String) {
    let reject = pendingReject
    pendingResolve = nil
    pendingReject = nil
    reject?(code, message, nil)
  }
}
